### PR TITLE
Enable nullable annotations

### PR DESCRIPTION
**Pull request by iamcarbon**
This PR:
- enables nullable annotations on the main library
- removes unused JetBrains annotations

### CODE REVIEW

**Comment by iamcarbon** [comment]
@EliotJones @BobLd Ready for feedback / review.

**Comment by BobLd** [comment]
@iamcarbon thanks a lot, I'll review it shortly

**Comment by BobLd** [comment]
@iamcarbon I've added comments. Thanks a lot for the work! As a side note, when working on very large PRs (many files to review), let's try to keep the focus of the PR on a single item (e.g. nullable). If you notice issue or bugs during your changes, I'm more than happy for you to create a separate PR to address these issues. Thanks

**Comment by iamcarbon** [comment]
@BobLd Let me know if you submitted your comments -- I don't see the on GitHub yet. I'll be careful on any future PR's to keep things tightly scoped to individual items.

**Review by BobLd — CHANGES_REQUESTED** [review]

**Comment by BobLd** [comment]
@iamcarbon my bad, they should be visible now

File: src/UglyToad.PdfPig/Polyfills/NullableAnnotations.cs
[the remark this answers is not in the document]
Comment (iamcarbon) [reply src/UglyToad.PdfPig/Polyfills/NullableAnnotations.cs:1]: A "Polyfill" conditionally patches an older framework to support newer framework features. It's been a more common pattern to back port modern .NET features to older features. It's possible we may want to add additional Polyfills for additional modern .NET features (e.g. to spanify streams), and these would be nicely contained here.

e.g. 

https://github.com/xoofx/markdig/tree/master/src/Markdig/Polyfills
https://www.meziantou.net/polyfills-in-dotnet-to-ease-multi-targeting.htm

[the remark this answers is not in the document]
Comment (BobLd) [reply src/UglyToad.PdfPig/Polyfills/NullableAnnotations.cs:1]: sounds good, I didn't know this naming convention. Thx for the explaination

File: src/UglyToad.PdfPig/PdfDocument.cs [hunk 1]
```diff
@@ -177,7 +167,7 @@ public Page GetPage(int pageNumber)
             }
             catch (Exception ex)
             {
-                if (IsEncrypted)
+                if (IsEncrypted && encryptionDictionary != null)
```
[the remark this answers is not in the document]
Comment (iamcarbon) [reply src/UglyToad.PdfPig/PdfDocument.cs:170]: encryptionDictionary is guaranteed to non-null when IsEncrypted is true. 

Adding this annotation made this extra check unnecessary. 

[MemberNotNullWhen(true, nameof(encryptionDictionary))]

File: src/UglyToad.PdfPig/Graphics/Colors/RGBWorkingSpace.cs
[the remark this answers is not in the document]
Comment (iamcarbon) [reply src/UglyToad.PdfPig/Graphics/Colors/RGBWorkingSpace.cs:1]: Without this change, these end up being errors with nullable annotations enabled.

We can, however, continue to always return false when passed a null color object to match the current behavior. 

Let me know if you have a preference.  

Screenshot 2024-03-17 at 11 09 21 AM

Screenshot 2024-03-17 at 11 09 29 AM

File: src/UglyToad.PdfPig/Tokenization/Scanner/PdfTokenScanner.cs [hunk 2]
```diff
@@ -703,7 +708,7 @@ public ObjectToken Get(IndirectReference reference)
 
             if (!objectLocationProvider.TryGetOffset(reference, out var offset))
             {
-                return null;
+                return null!;
```
[the remark this answers is not in the document]
Comment (iamcarbon) [reply src/UglyToad.PdfPig/Tokenization/Scanner/PdfTokenScanner.cs]: Very few of this methods callers check for null, and this resulted in a significant number  "!" suppressions.

For now, I opted this method out of nullability so we can improve further in the future.

As a better solution, I'm wondering if we split this function into two variants:

```
ObjectToken? Find(IndirectReference) -> returns null if not found
ObjectToken Get(IndirectReference) -> throws if not found

```

File: src/UglyToad.PdfPig/Content/Page.cs [hunk 3]
```diff
@@ -215,12 +213,12 @@ public IEnumerable<Annotation> GetAnnotations()
                 // TO DO
                 //var annots = GetAnnotations().ToList();
 
-                return mcesOptional.GroupBy(oc => oc.Name).ToDictionary(g => g.Key, g => g.ToList() as IReadOnlyList<OptionalContentGroupElement>);
+                return mcesOptional.GroupBy(oc => oc.Name).ToDictionary(g => g.Key!, g => (IReadOnlyList<OptionalContentGroupElement>)g.ToList());
```
[the remark this answers is not in the document]
Comment (iamcarbon) [reply src/UglyToad.PdfPig/Content/Page.cs:216]: Yep, this function continues to return a list. Moving the cast to the left just prevented this from being null.

**Comment by iamcarbon** [comment]
@BobLd Thanks for the review! Comments above.

File: src/UglyToad.PdfPig/Tokenization/Scanner/PdfTokenScanner.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (BobLd) [reply src/UglyToad.PdfPig/Tokenization/Scanner/PdfTokenScanner.cs]: okay I see, that makes sense. Let's leave it like that for the moment, and possibly implement your idea in a later PR

File: src/UglyToad.PdfPig/Graphics/Colors/RGBWorkingSpace.cs
[the remark this answers is not in the document]
Comment (BobLd) [reply src/UglyToad.PdfPig/Graphics/Colors/RGBWorkingSpace.cs:1]: This comment is refering `RGBWorkingSpace.cs`, not to `RGBColor.cs` as GitHub would make you believe

**Review by BobLd — CHANGES_REQUESTED** [review]

**Comment by BobLd** [comment]
Maybe it will be easier to post my last comment here: in the RGBWorkingSpace.cs files, I see you made some changes to the color / matrices. Do you mind reverting those and createing a different PR for that?

**Comment by BobLd** [comment]
Also do you mind adding the change to fix the nighlty build to this PR? I've created one here https://github.com/UglyToad/PdfPig/pull/804/files but that will be easier if it goes with yours - thx again!

File: src/UglyToad.PdfPig/Graphics/Colors/RGBWorkingSpace.cs
[the remark this answers is not in the document]
Comment (iamcarbon) [reply src/UglyToad.PdfPig/Graphics/Colors/RGBWorkingSpace.cs:1]: Oops! 

The constructor was added to the internal RGBWorkingSpace to ensure GammaCorrection was initialized as non null.

We can revert this change and suppress the nullability annotations on this file, if you prefer.

**Comment by iamcarbon** [comment]
> Also do you mind adding the change to fix the nighlty build to this PR? I've created one here https://github.com/UglyToad/PdfPig/pull/804/files but that will be easier if it goes with yours - thx again!

done!

File: src/UglyToad.PdfPig/Graphics/Colors/RGBWorkingSpace.cs
[the remark this answers is not in the document]
Comment (BobLd) [reply src/UglyToad.PdfPig/Graphics/Colors/RGBWorkingSpace.cs:1]: ok my bad - github was showing diffs where there were none:

image


All good thne

**Comment by BobLd** [comment]
okay I'm good - will be merged shortly

**Review by BobLd — APPROVED** [review]

**Pull request merged**